zone/expedition.cpp

```diff
@@ -0,0 +1,2275 @@
+			{
+				instance_ids.emplace_back(instance_id);
+			}
+
+			std::unique_ptr<Expedition> expedition = std::unique_ptr<Expedition>(new Expedition(
+				expedition_id,
+				row[col::uuid],                                         // expedition uuid
+				DynamicZone{ instance_id },
+				row[col::expedition_name],                              // expedition name
+				ExpeditionMember{ leader_id, row[col::leader_name] },   // expedition leader id, name
+				strtoul(row[col::min_players], nullptr, 10),            // min_players
+				strtoul(row[col::max_players], nullptr, 10)             // max_players
+			));
+
+			bool add_replay_on_join = (strtoul(row[col::add_replay_on_join], nullptr, 10) != 0);
+			bool is_locked = (strtoul(row[col::is_locked], nullptr, 10) != 0);
+
+			expedition->SetReplayLockoutOnMemberJoin(add_replay_on_join);
+			expedition->SetLocked(is_locked, ExpeditionLockMessage::None);
+
+			zone->expedition_cache.emplace(expedition_id, std::move(expedition));
+		}
+
+		last_expedition_id = expedition_id;
+
+		// looping expedition members
+		auto current_expedition = Expedition::FindCachedExpeditionByID(last_expedition_id);
+		if (current_expedition)
+		{
+			auto member_id = strtoul(row[col::member_id], nullptr, 10);
+			current_expedition->AddInternalMember(
+				row[col::member_name], member_id, ExpeditionMemberStatus::Offline);
+			expedition_character_ids.emplace_back(expedition_id, member_id);
+		}
+	}
+
+	// ask world for online members from all cached expeditions at once
+	Expedition::SendWorldGetOnlineMembers(expedition_character_ids);
+
+	// bulk load dynamic zone data and expedition lockouts for cached expeditions
+	auto dynamic_zones = DynamicZone::LoadMultipleDzFromDatabase(instance_ids);
+	auto expedition_lockouts = ExpeditionDatabase::LoadMultipleExpeditionLockouts(expedition_ids);
+
+	for (const auto& expedition_id : expedition_ids)
+	{
+		auto expedition = Expedition::FindCachedExpeditionByID(expedition_id);
+		if (expedition)
+		{
+			auto dz_iter = dynamic_zones.find(expedition->GetInstanceID());
+			if (dz_iter != dynamic_zones.end())
+			{
+				expedition->m_dynamiczone = dz_iter->second;
+			}
+
+			auto lockout_iter = expedition_lockouts.find(expedition->GetID());
+			if (lockout_iter != expedition_lockouts.end())
+			{
+				expedition->m_lockouts = lockout_iter->second;
+			}
+
+			// send member updates now that all data is loaded for the cached expedition(s)
+			expedition->SendUpdatesToZoneMembers();
+		}
+	}
+}
+
+void Expedition::CacheFromDatabase(uint32_t expedition_id)
+{
+	if (zone)
+	{
+		BenchTimer benchmark;
+
+		auto results = ExpeditionDatabase::LoadExpedition(expedition_id);
+		if (!results.Success())
+		{
+			LogExpeditions("Failed to load Expedition [{}] for zone cache", expedition_id);
+			return;
+		}
+
+		CacheExpeditions(results);
+
+		auto elapsed = benchmark.elapsed();
+		LogExpeditions("Caching new expedition [{}] took [{}s]", expedition_id, elapsed);
```

Comment:
This shouldn't need to be trace/detail level since it will only occur when a new expedition is created. It might be useful to leave this at normal logging so it can be more easily monitored to catch any performance issues (detail can get spammy even while I solo test)